Fix LXC container implementations

Nomsplease opened the pull request:
I moved branches around and it caused quite the Chaos on this PR. This a re-open of https://github.com/techno-tim/k3s-ansible/pull/219 with the same changes, just moved to a different branch

# Proposed Changes
<!--- Provide a general summary of your changes -->

- correct some implementation ideas for lxc supprt
- We should not assume the rc.local is empty and just overwrite it,  we can do this better and insert a managed code block
- We also should not assume we are root on reboot handler, and become has been added to this handler
- Also added resetting the lxc modification to a untouched state inside the container, resetting the lxc config in the host may also be a worthwhile adventure to make sure we get as close to original as possible.

## Checklist

- [x] Tested locally
- [x] Ran `site.yml` playbook
- [x] Ran `reset.yml` playbook
- [X] Did not add any unnecessary changes
- [X] Ran pre-commit install at least once before committing
- [X] 🚀

acdoussan commented on reset.yml:
```diff
@@ -13,3 +13,11 @@
+  remote_user: "{{ proxmox_lxc_ssh_user }}"
+  roles:
+    - role: reset_proxmox_lxc
+      when: proxmox_lxc_configure
```
Should this always run? I suppose there is a chance that the user configured these values outside of this playbook, and this would remove them

Nomsplease replied:
I feel like it should, only because I consider this playbook as a it should configure and manage everything. If we don't we could leave security issues where they don't need to exist without user knowledge.

acdoussan replied:
Think I agree, think the logic here should be that we assume these hosts are dedicated to k3s, so we should make sure this stuff is torn down through the reset process. Don't feel too strongly here though. Would just mean removing this when condition.

Nomsplease replied:
That or the user just does not specify that the LXC is configured and hits those functions manually. I personally do this as the LXCs are dedicated solely to K3s, as I imagine most people do.

acdoussan commented:
One super small comment, but I'm good with these changes as is 👍

timothystewart6 commented:
Looks like we have some lint.  We aren't using the literal `true` for become, we've been using `yes`

Nomsplease commented:
> Looks like we have some lint. We aren't using the literal `true` for become, we've been using `yes`

Im seeing in a lot of other files they are used interchangeably, but the standard for Ansible is True/False. Would you like me to swap these from True/False to Yes/No?

timothystewart6 commented:
More lint.  If install all of the dependencies it will fix and fail on commits, rather than later in CI

Nomsplease commented:
Ok that should be the last lint error, hopefully.

timothystewart6 commented:
Looks like the cleanup is failing

```log
 Remove rc.local modificaitons for proxmox lxc containers
```

timothystewart6 commented:
@Nomsplease  one small fix and then it's ready to go!

Nomsplease commented:
@timothystewart6 ill get this corrected and update when done. Did you run locally and get an error? Curious how that step failed for you.

timothystewart6 commented:
@Nomsplease  you can see it failing in CI at this line
https://github.com/techno-tim/k3s-ansible/actions/runs/4178688825/jobs/7313361925#step:9:966

timothystewart6 commented:
(expand and look at line 966)

Nomsplease commented:
Updated branch and changed how we handle the rc.local file. This is a good catch because we should not be touching this file on a host that was not setup with lxc in the first place.

sleiner commented on roles/lxc/tasks/main.yml:
```diff
@@ -1,7 +1,21 @@
     mode: "u=rwx,g=rx,o=rx"
+  when: not rcfile.stat.exists
+
+- name: Wrtie rc.local file
```
```suggestion
- name: Write rc.local file
```

sleiner commented on roles/reset/tasks/main.yml:
```diff
@@ -54,3 +54,31 @@
+    path: /etc/rc.local
+  register: rcfile
+
+- name: Remove rc.local modificaitons for proxmox lxc containers
```
```suggestion
- name: Remove rc.local modifications for proxmox lxc containers
```

sleiner commented on roles/reset_proxmox_lxc/tasks/main.yml:
```diff
@@ -0,0 +1,53 @@
+  loop: "{{ proxmox_lxc_ct_ids }}"
+  register: stat_results
+
+- name: filter out files that do not exist
+  set_fact:
+    proxmox_lxc_filtered_files:
+      '{{ stat_results.results | rejectattr("stat.exists", "false") | map(attribute="stat.path") }}'
+
+# used for the reboot handler
+- name: get container ids from filtered files
+  set_fact:
+    proxmox_lxc_filtered_ids:
+      '{{ proxmox_lxc_filtered_files | map("split", "/") | map("last") | map("split", ".") | map("first") }}'
+
```
for the sake of shorter logs for users: Can we merge this into a single task?

acdoussan replied:
No, both of these variables are used.

Also, these filters are already pretty relatively complicated, I think the readability of having them separate is worth having one extra task, even if the variable wasn't needed.

sleiner replied:
> No, both of these variables are used.

Obviously :-) What are you getting at?

> readability of having them separate

What do you think of this?

```suggestion
- name: Build list of files and container IDs
  vars:
    existing_files: >
      {{ stat_results.results | rejectattr("stat.exists", "false") | map(attribute="stat.path") }}
  set_fact:
    proxmox_lxc_filtered_files: >
      {{ existing_files }}
    proxmox_lxc_filtered_ids: >
      {{ existing_files | map("split", "/") | map("last") | map("split", ".") | map("first") }}
```

same logic, no bigger filter expressions than before, still more compact to developers as well as users of the playbook 👍🏻

acdoussan replied:
Oh, think I just misunderstood, gotcha yeah, indifferent here. Don't need that existing_files var though no? could just have that  set directly on proxmox_lxc_filtered_files and then use that in proxmox_lxc_filtered_ids

sleiner replied:
> Don't need that existing_files var though no?

I'm pretty sure that the variable is needed (since this is not imperative programming, and as a coarse model, one could say that both of the facts are set at the same time). For example, this playbook cannot be run:

```yaml
- hosts: localhost
  tasks:
    - ansible.builtin.set_fact:  # You will encounter an error while executing this task
        foo: "42"
        bar: "{{ foo }}"

    - ansible.builtin.debug:
        msg: foo={{foo}}, bar={{bar}}
```

acdoussan replied:
gotcha, alright, if we do this I think it would be worth it to update the implementation as well, rather than just the reset task. Probably best as its own PR, this one has been open for quite a while now.

sleiner replied:
Feel free to adapt the implementation as well, I'm all for it 👍🏻
As for this specific snippet: If we agree that this is the pattern we should use, the patch is ready and one would just click a button to apply it, I see no reason to merge it with the old pattern - do you?

acdoussan replied:
I'll leave that up to @Nomsplease and if they want to do another revision and round of testing to make sure there is nothing we haven't considered :)

Nomsplease replied:
I like this idea, but I feel like we are getting a little into the weeds on this. 

The current system functions and I think is more clear to a newer Anisble user, and we have already done all the validation testing on the massive amount of changes in the PR.

I would say if this is still desired to open another PR to adjust these filter functions, I do not intended to implement this change.

acdoussan replied:
It is more than CI, this role isn't covered by the molecule tests (for reasons that can be seen in this [PR](https://github.com/techno-tim/k3s-ansible/pull/209)), you have to manually create/destroy a new cluster with lxc containers to test this change.

sleiner commented on roles/reset/tasks/main.yml:
```diff
@@ -54,3 +54,31 @@
   file:
     path: /tmp/k3s
     state: absent
+
+- name: Check if rc.local exists
+  stat:
+    path: /etc/rc.local
+  register: rcfile
```
Can you explain how the LXC specific tasks in the `reset` role are different from the tasks in the `reset_proxmox_lxc` role?

acdoussan replied:
reset tasks run on the actual container running k3s (the lxc containers) reset_proxmox_lxc runs on the proxmox hosts directly.

sleiner replied:
I see 👍🏻

Nomsplease commented:
> We're so close to the finish line. What's left?

All we had were the other review changes, but I think we are good to go on the functionality that was intended in this change.

Nomsplease commented:
I pulled the current upstream into this branch. One last round of checks should not hurt here and we can make the merge as smooth as possible.

timothystewart6 commented:
Thank you all for working through this.  If we need to change a pattern, we can add a new issue for it!  Thanks again everyone!
